include/networkit/graph/Graph.hpp

```diff
@@ -139,6 +145,389 @@ class Graph final {
+            other.valid = false;
+            if (valid)
+                owned_storage->attrSet.insert(this);
+        }
+
+        ~NodeAttribute() {
+            if (owned_storage) {
+                owned_storage->attrSet.erase(this);
+            }
+        }
+
+        auto begin() { return Iterator(owned_storage.get()).nextValid(); }
+
+        auto end() { return Iterator(nullptr); }
+
+        auto size() { return owned_storage->size(); }
+
+        void set(index i, T v) {
+            checkAttribute();
+            owned_storage->set(i, std::move(v));
+        }
+
+        auto get(index i) {
+            checkAttribute();
+            return owned_storage->get(i);
+        }
+
+        auto get(index i, T defaultT) {
+            checkAttribute();
+            return owned_storage->get(i, defaultT);
+        }
+
+        IndexProxy operator[](index i) {
+            checkAttribute();
+            return IndexProxy(owned_storage.get(), i);
+        }
+
+        void checkAttribute() {
+            if (!valid) {
+                throw std::runtime_error("Invalid attribute");
+            }
+        }
+
+        void write(const std::string &filename) {
+            std::ofstream out(filename);
+            if (!out)
+                std::cerr << "cannot open '" << filename << "' for writing\n";
+
+            for (auto it = begin(); it != end(); ++it) {
+                auto pair = *it;
+                auto n = pair.first;
+                auto v = pair.second;
+                out << n << "\t" << v << "\n";
+            }
+            out.close();
+        }
+
+        void read(const std::string &filename) {
+            std::ifstream in(filename);
+            if (!in) {
+                std::cerr << "cannot open '" << filename << "' for reading\n";
+            }
+            while (true) {
+                int n;
+                T t;
+                std::string line;
+                std::getline(in, line);
+                if (!in)
+                    break;
+                std::istringstream istring(line);
+                istring >> n >> t;
+                std::cout << "got: " << n << " " << t << "\n";
+            }
+        }
+
+    private:
+        void invalidateAttribute() { valid = false; }
+
+    private:
+        std::shared_ptr<NodeAttributeStorage<T>> owned_storage;
+        bool valid;
+        friend NodeAttributeStorage<T>;
+    }; // class NodeAttribute
+
+    class NodeAttributeMap {
+        friend Graph;
+        const Graph *theGraph;
+
+    public:
+        std::unordered_map<std::string, std::shared_ptr<NodeAttributeStorageBase>> attrMap;
+
+        NodeAttributeMap(const Graph *g) : theGraph{g} {}
+
+        auto find(std::string const &name) {
+            auto it = attrMap.find(name);
+            if (it == attrMap.end()) {
+                throw std::runtime_error("No such attribute");
+            }
+            return it;
```

Comment:
This can be simplified with `return attrMap.at(name);`

Reply:
But the error message is a bit more helpful this way.